shell/platform/android/external_view_embedder/surface_pool.cc

```diff
@@ -0,0 +1,75 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/android/external_view_embedder/surface_pool.h"
+
+namespace flutter {
+
+OverlayLayer::OverlayLayer(int id,
+                           std::unique_ptr<AndroidSurface> android_surface,
+                           std::unique_ptr<Surface> surface)
+    : id(id),
+      android_surface(std::move(android_surface)),
+      surface(std::move(surface)){};
+
+OverlayLayer::~OverlayLayer() = default;
+
+SurfacePool::SurfacePool() = default;
+
+SurfacePool::~SurfacePool() = default;
+
+std::shared_ptr<OverlayLayer> SurfacePool::GetLayer(
```

Comment:
Once created, the overlay will never be collected no matter how long it goes unused. What you want is something like the following:

1: Ask the pool for all the layers you need. 
2: After you are done asking, remove any unused layers from the pool.
3: Use the layers you just got for frame operations.
4: Give the layers back to the surface pool.

That way, any layers that you don't reuse the next frame will get collected.

Reply:
This technique is conceptually similar to the one used for the external view embedders (see [`flutter::EmbedderRenderTargetCache`](https://github.com/flutter/engine/blob/master/shell/platform/embedder/embedder_render_target_cache.h)), there the order is the following:

1: One or many calls to `EmbedderRenderTargetCache::GetExistingTargetsInCache`. The caller creates the render target if it doesn't exist unlike in this implementation where the pool creates it. Either way is fine.
2: Caller scribbles in the render targets.
3: Caller calls `ClearAllRenderTargetsInCache`.
4: For all the render targets owned by the caller, ownership is transferred back to the "pool" via `CacheRenderTarget`

Reply:
This one is tricky. On one hand, this is by design as building a heuristic for deallocating overlay surfaces is possible, but it requires some research. For example, determining when is long enough so these surfaces can be deallocated. 

The issue is that we want to reuse as many surfaces as possible between frames as creating new ones have perf implications.

Potentially, the technique used for thread merging (extend a lease) could potentially be used here. cc @cyanglaz

Reply:
I filed https://github.com/flutter/flutter/issues/59904